Implemented BLEU score, wrote unit tests and documentation for it.

kadamrahul18 opened the pull request:
### **BLEU Metric Implementation:**

1. Added a new BLEU class under sdks/python/src/opik/evaluation/metrics/heuristics/bleu.py.
2. Implemented the BLEU algorithm to calculate scores based on n-gram precision between the generated text and a reference text.
3. Included methods for handling both single sentences and corpus-level scoring.
4. Implemented smoothing techniques (methods 0, 1, 2, 3 from the Chen & Cherry paper) to address zero n-gram matches.
5. Added configuration options for n-gram order, smoothing method, and weights.

**Unit Tests:**

1. Added comprehensive unit tests in sdks/python/tests/unit/evaluation/metrics/test_heuristics.py to validate the BLEU metric's behavior in various scenarios:

- Exact match, partial match, and no match cases.
- Empty candidate and reference strings.
- Different smoothing methods.
- Corpus-level scoring.
- Edge cases and error handling.

**Integration with Evaluation Framework:**

1. Added the BLEU class to the __all__ list in sdks/python/src/opik/evaluation/metrics/heuristics/__init__.py to make it discoverable by the evaluate function.

**Documentation:**

1. Added a new documentation page for the BLEU metric (bleu.md) in the evaluation/metrics section of the documentation, detailing its purpose and usage.

**Testing:**

1. Thorough unit tests have been included to cover different aspects of the BLEU metric implementation, including edge cases and different smoothing methods.
2. All tests in the Python SDK, including the new tests for the BLEU metric, pass successfully when running pytest tests/ from the sdks/python directory.
3. pre-commit run --all-files has been executed successfully from the sdks/python directory, ensuring code style and formatting consistency.

**Request for Review:**

Please review the following aspects of this pull request:

1. Correctness of the BLEU metric implementation, including n-gram precision, brevity penalty, and smoothing.
2. Clarity and completeness of the unit tests.
3. Thoroughness of the documentation.
4. Adherence to Opik's coding standards and best practices.

Any feedback or suggestions for improvement are greatly appreciated.

alexkuzmik commented:
Hi @kadamrahul18!
I can see that the code is based on the `nltk` library implementation, which is one of the most popular libraries when it comes to BLEU score calculation.
I prefer to just use NLTK as well, it's likely not the last heuristic metric we will add, and I don't think that populating the code base with non-trivial mathematical calculations is the right thing to do when there are already pretty stable and specialized tools for that.
What I suggest doing is something like that:
```python
try:
    import nltk  # we won't add nltk as a package dependency, but we can add it to a separate requirements file for unit tests
except ImportError:
    nltk = None

...
class BLEU:
    def __init__(...):
        if nltk is None:
            raise ImportError("`nltk` library is required for BLEU score calculation, please install it via `pip install nltk`")
```
Under the hood of the metric implementation you can use `nltk.translate.bleu_score.sentence_bleu` or `nltk.translate.bleu_score.corpus_bleu`. 

That way we'll be able to have a stable implementation and avoid a big chunk of mathematical code (which is almost always hard to read and easy to break :) )

kadamrahul18 commented:
Hi @alexkuzmik, thanks for the feedback and suggestion! I've updated the code to use nltk's implementation instead of a custom one.

Please take a look at the updated code and let me know if you have any further suggestions or if anything is unclear. I appreciate your thorough review and helpful recommendations!

alexkuzmik commented on sdks/python/src/opik/evaluation/metrics/heuristics/bleu.py:
```diff
@@ -0,0 +1,218 @@
+        """
+        return getattr(self._nltk_smoother, self.smoothing_method, self._nltk_smoother.method0)
+
+    def _truncate_weights(self, candidate_len: int) -> tuple:
```
return type should be more precise, e.g. `Tuple[float, ...]`

alexkuzmik commented on sdks/python/src/opik/evaluation/metrics/heuristics/bleu.py:
```diff
@@ -0,0 +1,218 @@
+        :param name: Name for this metric instance.
+        :param track: Whether or not this metric is tracked (depends on your system).
+        :param n_grams: Up to which n-gram order to use (1 through n_grams).
+        :param smoothing_method: One of NLTK's SmoothingFunction methods (e.g., "method0", "method1", "method2", etc.).
```
Let's add a link in the docstrings https://www.nltk.org/api/nltk.translate.bleu_score.html#nltk.translate.bleu_score.SmoothingFunction
so that people could understand those options easier.

alexkuzmik commented on sdks/python/src/opik/evaluation/metrics/heuristics/bleu.py:
```diff
@@ -0,0 +1,218 @@
+
+try:
+    import nltk
+    from nltk.translate.bleu_score import sentence_bleu, corpus_bleu, SmoothingFunction
```
In order to keep the namespace clean and readable, we **import modules, not names**.
So that this code needs to be turned into smth like
```python
from nltk.translate import bleu_score as nltk_bleu_score

# and then later use the required API like that
nltk_bleu_score.SmoothingFunction
nltk_bleu_score.sentence_bleu
nltk_bleu_score.corpus_bleu
```
It is usually significantly easier to read the code when it explicitly tells where is the function or the class from :)

alexkuzmik commented on sdks/python/src/opik/evaluation/metrics/heuristics/bleu.py:
```diff
@@ -0,0 +1,218 @@
+        """
+        # 1) Handle empty candidate
+        if not output.strip():
+            return score_result.ScoreResult(
```
I think in such case we should raise `opik.exceptions.MetricComputationError` (you can see the examples in other metrics). Otherwise the code will silently "calculate" values when the user accidentally provided invalid inputs.

This comment is relevant for all other places where incorrect inputs just lead to zero value scores. We want people to know if they did something wrong instead of investigating low metric values :) 

you can take a look at opik.evaluation.scorer._score_test_case and see that it handles opik.exceptions.MetricComputationError and marks ScoreResult as failed.

kadamrahul18 replied:
Thank you for the feedback! I noticed that MetricComputationError isn’t currently defined in exceptions.py. Since we do need a specific exception for handling invalid metric inputs, I’ll create a new class for it there (e.g.:

```python
class MetricComputationError(OpikException):
    pass
```
   
Let me know if this approach works for you or if there’s a preferred existing exception I should use instead.

alexkuzmik commented on sdks/python/src/opik/evaluation/metrics/heuristics/bleu.py:
```diff
@@ -0,0 +1,218 @@
+        weights: Optional[List[float]] = None,
+    ):
+        """
+        :param name: Name for this metric instance.
```
Please use the same docstrings format as in our other metrics.

alexkuzmik commented on sdks/python/src/opik/evaluation/metrics/heuristics/bleu.py:
```diff
@@ -0,0 +1,218 @@
+        normalized = [w / w_sum for w in used_weights]
+        return tuple(normalized)
+
+    ###########################################################################
```
No need for these comments

alexkuzmik commented on sdks/python/src/opik/evaluation/metrics/heuristics/bleu.py:
```diff
@@ -0,0 +1,218 @@
+            else:
+                refs = []
+                skip_this = False
+                for r in ref_item:
```
I'd suggest improve naming and don't use such shortcuts, it affects readability.
`ref_item` doesn't tell that it's a collection of some values, it already looks like an item in collection.
Then what is `r` in `ref_item`? ref item item? :)

alexkuzmik commented on sdks/python/src/opik/evaluation/metrics/heuristics/bleu.py:
```diff
@@ -0,0 +1,218 @@
+            )
+
+        # Determine the largest candidate length
+        max_len = max(len(c) for c in all_candidates)
```
c -> candidate

alexkuzmik commented on sdks/python/tests/unit/evaluation/metrics/test_heuristics.py:
```diff
@@ -1,9 +1,11 @@
+import pytest
```
don't forget to add `nltk` to `tests/unit/test_requirements.txt`, or these unit tests won't work in our CI

alexkuzmik commented on sdks/python/src/opik/evaluation/metrics/heuristics/bleu.py:
```diff
@@ -0,0 +1,218 @@
+
+        self._nltk_smoother = SmoothingFunction()
+
+    def _get_smoothing_func(self):
```
return type is missing

alexkuzmik commented on sdks/python/src/opik/evaluation/metrics/heuristics/bleu.py:
```diff
@@ -0,0 +1,218 @@
+        self, output: str, reference: Union[str, List[str]], **ignored_kwargs: Any
+    ) -> score_result.ScoreResult:
+        """
+        Computes a single-sentence BLEU score using nltk.translate.bleu_score.sentence_bleu.
```
It's a public api method, please follow the same docstring format as in our other metrics.

alexkuzmik commented on sdks/python/src/opik/evaluation/metrics/heuristics/bleu.py:
```diff
@@ -0,0 +1,218 @@
+                reason="Mismatch: number of candidates != number of references.",
+            )
+
+        all_candidates = []
```
Please add type hints to these lists. It might be very helpful to understand the logic below (because of the fact that sometimes some values can be lists of strings and sometimes just strings).

alexkuzmik reviewed (changes requested):
Nice, thank you @kadamrahul18!
After delegating most of the algorithm part to `nltk` it's now significantly easier to review the PR :)
I left my comments.

kadamrahul18 commented:
Hi @alexkuzmik ,

Thank you so much for your detailed and insightful feedback! I really appreciate you taking the time to review the code so thoroughly.

I've implemented all of your suggestions. I've also addressed other minor suggestions and style issues you pointed out.

Please take another look and let me know if you have any further feedback. I'm happy to make any additional changes needed.

Thanks again for your valuable input!

kadamrahul18 commented:
Hi @alexkuzmik,

You are absolutely right, my apologies for that! I should have split BLEU into SentenceBLEU and CorpusBLEU in the previous iteration as you suggested.

I believe this addresses all the points from the previous review round. Please let me know if you have any further feedback!

alexkuzmik commented on sdks/python/src/opik/evaluation/metrics/heuristics/bleu.py:
```diff
@@ -0,0 +1,279 @@
+
+        self._smoother = nltk_bleu_score.SmoothingFunction()
+
+    def _get_smoothing_func(self) -> nltk_bleu_score.SmoothingFunction:
```
Looks good, thank you!

Only one thing after I ran the tests, we need to rename
`nltk_bleu_score.SmoothingFunction -> "nltk_bleu_score.SmoothingFunction"`
That's why the e2e tests are failing, they don't have `nltk` package dependency and can't import this module.

You can ignore the rest of the errors in CI except for the unit tests and e2e tests, other workflows are red just just because they require different API keys.

alexkuzmik commented:
@kadamrahul18 thank you for your work! I'm merging this to main branch and it will be included in the next SDK release :)
